packages/askar/src/wallet/AskarWallet.ts

```diff
@@ -0,0 +1,745 @@
+
+    for (const recipient of recipients) {
+      let recipientKeyEntry
+      try {
+        recipientKeyEntry = await this.session.fetchKey({ name: recipient.kid })
+      } catch (error) {
+        // TODO: Currently Askar wrapper throws error when key is not found
+        // In this case we don't need to throw any error because we should
+        // try with other recipient keys
+        continue
+      }
+      if (recipientKeyEntry) {
+        const recip_x = recipientKeyEntry.key.convertkey({ algorithm: KeyAlgs.X25519 })
+        recipientKey = recipient.kid
+
+        if (recipient.sender && recipient.iv) {
+          senderKey = new TextDecoder().decode(
+            CryptoBox.sealOpen({
+              recipientKey: recip_x,
+              ciphertext: recipient.sender,
+            })
+          )
+          const sender_x = AskarKey.fromPublicBytes({
+            algorithm: KeyAlgs.Ed25519,
+            publicKey: decodeFromBase58(senderKey),
+          }).convertkey({ algorithm: KeyAlgs.X25519 })
+
+          payloadKey = CryptoBox.open({
+            recipientKey: recip_x,
+            senderKey: sender_x,
+            message: recipient.encrypted_key,
+            nonce: recipient.iv,
+          })
+        }
+        break
+      }
+    }
+    if (!payloadKey) {
+      throw new WalletError('No corresponding recipient key found')
+    }
+
+    if (!senderKey && isAuthcrypt) {
+      throw new WalletError('Sender public key not provided for Authcrypt')
+    }
+
+    const cek = AskarKey.fromSecretBytes({ algorithm: KeyAlgs.Chacha20C20P, secretKey: payloadKey })
+    const message = cek.aeadDecrypt({
+      ciphertext: new Uint8Array(Buffer.from(messagePackage.ciphertext as any, 'base64')),
+      nonce: new Uint8Array(Buffer.from(messagePackage.iv as any, 'base64')),
+      tag: new Uint8Array(Buffer.from(messagePackage.tag as any, 'base64')),
+      aad: new TextEncoder().encode(messagePackage.protected),
+    })
+
+    return {
+      plaintextMessage: JsonEncoder.fromBuffer(message),
+      senderKey,
+      recipientKey,
+    }
+  }
+
+  public async generateNonce(): Promise<string> {
+    try {
+      return new TextDecoder().decode(CryptoBox.randomNonce())
+    } catch (error) {
+      if (!isError(error)) {
+        throw new AriesFrameworkError('Attempted to throw error, but it was not of type Error', { cause: error })
+      }
+      throw new WalletError('Error generating nonce', { cause: error })
+    }
+  }
+
+  public async generateWalletKey() {
+    try {
+      return Store.generateRawKey()
+    } catch (error) {
+      throw new WalletError('Error generating wallet key', { cause: error })
+    }
+  }
+
+  private async getAskarWalletConfig(walletConfig: WalletConfig) {
+    const { uri, path } = uriFromWalletConfig(walletConfig, this.fileSystem.basePath)
+
+    // Make sure path exists before creating the wallet
+    if (path) {
+      await this.fileSystem.createDirectory(path)
+    }
```

Comment:
Will it throw an error if the path doesn't exist?

We need to take a good look at the fileSystem basePath. I think for some reason the fileSystem by default is a temp directory. See https://github.com/hyperledger/aries-framework-javascript/issues/706

Reply:
Yes, unfortunately that's what I've found for SQLite backend. In the code I left the possibility of choosing our very own path or just leave the framework to decide (based on that `basePath` that I agree we should not put on a temporary directory). This could be greatly improved as soon as we define a better wallet config interface for Askar.